File: 2482. Difference Between Ones and Zeros in Row and Column.py
class Solution:
    # solution 2: in-place grid modification with a trick!
    # O(m * n) time and O(m + n) space complexity
    def onesMinusZeros(self, grid: List[List[int]]) -> List[List[int]]:
        m, n = len(grid), len(grid[0])
        r, c = [0] * m, [0] * n

        for i, row in enumerate(grid):
            for j, num in enumerate(row):
                r[i] += num
                c[j] += num

        for i in range(m):
            for j in range(n):
                grid[i][j] = 2 * (r[i] + c[j]) - (m + n)

        return grid

    # # solution 1
    # # O(m * n) time and space complexity
    # def onesMinusZeros(self, grid: List[List[int]]) -> List[List[int]]:
    #     r = {
    #         0: [0] * len(grid),
    #         1: [0] * len(grid)
    #     }
    #     c = {
    #         0: [0] * len(grid[0]),
    #         1: [0] * len(grid[0])
    #     }

    #     for i, row in enumerate(grid):
    #         for j, num in enumerate(row):
    #             r[num][i] += 1
    #             c[num][j] += 1

    #     diff = []
    #     for i in range(len(grid)):
    #         diff.append([])
    #         for j in range(len(grid[0])):
    #             zeroes = r[0][i] + c[0][j]
    #             ones = r[1][i] + c[1][j]
    #             diff[i].append(ones - zeroes)

    #     return diff

File: 2482. Difference Between Ones and Zeros in Row and Column.cpp
class Solution {
public:
    // Solution 2: in-place grid manipulation, with a trick!
    // O(m * n) time and O(m + n) space complexity 
    vector<vector<int>> onesMinusZeros(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        vector<int> r(m);
        vector<int> c(n);

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                r[i] += grid[i][j];
                c[j] += grid[i][j];
            };
        };

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                grid[i][j] = 2 * (r[i] + c[j]) - (m + n);
            };
        };

        return grid;
    };

    // // solution 1
    // // O(m * n) time and space complexity 
    // vector<vector<int>> onesMinusZeros(vector<vector<int>>& grid) {
    //     map<int, vector<int>> r;
    //     r[0] = {};
    //     r[1] = {};
    //     for (int i = 0; i < grid.size(); i++) {
    //         r[0].push_back(0);
    //         r[1].push_back(0);
    //     };
        
    //     map<int, vector<int>> c;
    //     c[0] = {};
    //     c[1] = {};
    //     for (int i = 0; i < grid[0].size(); i++) {
    //         c[0].push_back(0);
    //         c[1].push_back(0);
    //     };

    //     for (int i = 0; i < grid.size(); i++) {
    //         for (int j = 0; j < grid[0].size(); j++) {
    //             int num = grid[i][j];
    //             r[num][i]++;
    //             c[num][j]++;
    //         };
    //     };

    //     vector<vector<int>> diff(grid.size());
    //     for (int i = 0; i < grid.size(); i++) {
    //         for (int j = 0; j < grid[0].size(); j++) {
    //             int zeroes = r[0][i] + c[0][j];
    //             int ones = r[1][i] + c[1][j];
    //             diff[i].push_back(ones - zeroes);
    //         };
    //     };
        
    //     return diff;
    // };
};
